File: pa4/List.c
//-----------------------------------------------------------------------------
// Spenser Estrada 
// spmestra
// pa4
//
// List.c
// Implementation file for List ADT
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// structs --------------------------------------------------------------------

// private NodeObj type
typedef struct NodeObj{
   struct NodeObj* prev;
   int index;
   int data;
   struct NodeObj* next;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private ListObj type
typedef struct ListObj{
   Node front;
   Node back;
   int length;
   Node cursor;
} ListObj;


// Constructors-Destructors ---------------------------------------------------

// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data){
   Node N = malloc(sizeof(NodeObj));
   N->prev = NULL;
   N->data = data;
   N->next = NULL;
   return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

// newList()
// Returns reference to new empty List object.
List newList(void){
   List L;
   L = malloc(sizeof(ListObj));
   L->front = L->back = NULL;
   L->length = 0;
   L-> cursor = NULL;
   return(L);
}

// freeList()
// Frees all heap memory associated with List *pL, and sets *pL to NULL.S
void freeList(List* pL){
   if(pL!=NULL && *pL!=NULL) { 
      while( length(*pL)!=0 ) { 
         deleteBack(*pL); 
      }
      free(*pL);
      *pL = NULL;
   }
}


// Access functions -----------------------------------------------------------

// length()
// Returns the length of L.
int length(List L){
   if( L==NULL ){
      printf("List Error: calling length() on NULL List reference\n");
      exit(1);
   }
   return(L->length);
}

// getIndex()
// Returns the index of the cursor element in this list, or returns -1 if the
// cursor element is undefined.
int getIndex(List L){
   if (L->cursor == NULL) return -1;
   return L->cursor->index;
}

// front()
// Returns the value at the front of L.
// Pre: length(L)>0
int front(List L){
   if( L==NULL ){
      printf("List Error: calling front() on NULL List reference\n");
      exit(1);
   }
   if( length(L)==0 ){
      printf("List Error: calling front() on an empty List\n");
      exit(1);
   }
   return(L->front->data);
}

// back()
// Returns the value at the back of L.
// Pre: length(L)>0
int back(List L){
   if( L==NULL ){
      printf("List Error: calling back() on NULL List reference\n");
      exit(1);
   }
   if( length(L)==0 ){
      printf("List Error: calling back() on an empty List\n");
      exit(1);
   }  
   return(L->back->data);
}  

// getElement()
// Returns cursor element in this list
// Pre: length(L)>0, getIndex(L) >=0
int getElement(List L){
   if( L==NULL ){
      printf("List Error: calling getElement() on NULL List reference\n");
      exit(1);
   }
   if( length(L)==0 ){
      printf("List Error: calling getElement() on an empty List\n");
      exit(1);
   }
   if( getIndex(L)<0 ){
      printf("List Error: calling getElement() with no item selected\n");
      exit(1);
   }
   return L->cursor->data;
}

// equals()
// returns true (1) if A is identical to B, false (0) otherwise
int equals(List A, List B){
   int flag = 1;
   Node N = NULL;
   Node M = NULL;

   if( A==NULL || B==NULL ){
      printf("List Error: calling equals() on NULL List reference\n");
      exit(1);
   }
   N = A->front;
   M = B->front;
   if( A->length != B->length ) { return 0; }
   while( flag && N!=NULL){
      flag = (N->data==M->data);
      N = N->next;
      M = M->next;
   }
   return flag;
}


// Manipulation procedures ----------------------------------------------------

// clear()
// Resets List to the empty state.
void clear(List L){
   if( L==NULL ){
      printf("List Error: calling clear() on NULL List reference\n");
      exit(1);
   }
   Node temp;
   while ( L->front!=NULL ) {
      temp= L->front->next;
      freeNode( &L->front );
      L->front = temp;
   }
   L->front = L->back = L->cursor = NULL;
   L->length=0;
}

// moveTo()
// If 0<=i<=length(L)-1, moves cursor to element at index i, otherwise, cursor
// becomes undefined.
void moveTo(List L, int i){
   if( L==NULL ){
      printf("List Error: calling moveTo() on NULL List reference\n");
      exit(1);
   } 
   if (getIndex(L) != i){
      if ( 0 <= i && i <= ( length(L)-1) ){ 
         if( getIndex(L) <0 ){
            L->cursor = L->front;
            while ( getIndex(L) != i ){
               L->cursor = L->cursor->next;
            }
         } 
         if( getIndex(L) < i ){
            while ( getIndex(L) != i ){
               L->cursor = L->cursor->next;
            }
         }
         if( getIndex(L) > i ){
            while( getIndex(L) != i ){
               L->cursor = L->cursor->prev;
            }
         }
      }
   }else L->cursor = NULL;
}

// movePrev()
// If 0<getIndex(L)<=length(L)-1, moves the cursor one step toward the front 
// of the list. If getIndex(L)==0, cursor becomes undefined.
// If getIndex(L)==-1, cursor remains undefined. This operation is equivalent
// to moveTo(getIndex(L)-1). 
void movePrev(List L){
   if( L==NULL ){
      printf("List Error: calling movePrev() on NULL List reference\n");
      exit(1);
   }
   if ( 0 < getIndex(L) && getIndex(L) <= length(L)-1 ){
      L->cursor = L->cursor->prev;
   } else L->cursor = NULL;
}


// moveNext()
// If 0<=getIndex(L)<length(L)-1, moves the cursor one step toward the back
// of the list. If getIndex(L)==length(L)-1, cursor becomes undefined. 
// If index==-1, cursor remains undefined. This operation is equivalent 
// to moveTo(getIndex(L)+1). 
void moveNext (List L) {
   if( L==NULL ){
      printf("List Error: calling moveNext() on NULL List reference\n");
      exit(1);
   }
   if( 0 <= getIndex(L) && getIndex(L) < length(L)-1 ){
      L->cursor = L->cursor->next;
   }else L->cursor = NULL;
}

// prepend()
// Places data element at the front of L
void prepend(List L, int data)
{
   Node N = newNode(data);

   if( L==NULL ){
      printf("List Error: calling prepend() on NULL List reference\n");
      exit(1);
   }
   if( length(L)==0 ){ 
      N->index = 0;
      L->front = L->back = N; 
   }
   else { 
      L->front->prev = N; N->next=L->front; L->front = N;
      int i = 0;
      Node temp = L->front;
      while (temp != NULL){
         temp->index = i;
         i++;
         temp=temp->next;
      }
   }
   L->length++;
}

// append()
// Places new data element at the end of L
void append(List L, int data)
{
   Node N = newNode(data);

   if( L==NULL ){
      printf("List Error: calling append() on NULL List reference\n");
      exit(1);
   }
   if( length(L)==0 ) { 
      N->index = 0;
      L->front = L->back = N; 
   }
   else{ 
      N->index = (L->back->index +1);
      L->back->next = N; N->prev = L->back; L->back = N; 
   }
   L->length++;
}

// insertBefore()
// Inserts new element before cursor element in this List.
// Pre : length(L)>0, getIndex(L)>=0 
void insertBefore(List L, int data){
   Node N = newNode(data);
   
   if( L==NULL ){
      printf("List Error: calling insertBefore() on NULL List reference\n");
      exit(1);
   }
   if( length(L)==0 ){
      printf("List Error: calling insertBefore() on empty List\n");
      exit(1);
   }
   if( getIndex(L)<0 ){
      printf("List Error: calling insertBefore() with undefined cursor\n");
      exit(1);
   }
   if (L->cursor == L->front ){
      prepend(L, data);
   }else{
      L->cursor->prev->next = N; N->prev = L->cursor->prev;
      N->next = L->cursor; L->cursor->prev = N;
      int i = 0; 
      Node temp = L->front;
      while (temp != NULL){
         temp->index = i;
         i++;
         temp=temp->next;
      }
   }
   L->length++;
}

// insertAfter()
// Inserts new element after cursor element in this List.
// Pre: length(L)>0, getIndex(L)>=0
void insertAfter(List L, int data) {
   Node N = newNode(data);

   if( L==NULL ){
      printf("List Error: calling insertAfter() on NULL List reference\n");
      exit(1);
   }
   if( length(L)==0 ){
      printf("List Error: calling insertAfter() on empty List\n");
      exit(1);
   }
   if( getIndex(L)<0 ){
      printf("List Error: calling insertAfter() with undefined cursor\n");
      exit(1);
   }
   if (L->cursor == L->back ){
      append(L, data);
   }else{
      L->cursor->next->prev = N; N->next = L->cursor->next;
      N->prev = L->cursor; L->cursor->next = N;
      int i = 0;
      Node temp = L->front;
      while (temp != NULL){
         temp->index = i;
         i++;
         temp=temp->next;
      }
   }
   L->length++;
}

// deleteFront()
// Deletes element at front of L
// Pre: length(L) !=0
void deleteFront(List L){
   Node N = NULL;

   if( L==NULL ){
      printf("List Error: calling deleteFront() on NULL List reference\n");
      exit(1);
   }
   if( length(L)==0 ) {
      printf("List Error: calling deleteFront() on an empty List\n");
      exit(1);
   }
   N = L->front;
   if( length(L)>1 ) { L->front = L->front->next; L->front->prev=NULL; }
   else { L->front = L->back = NULL; }
   L->length--;
   freeNode(&N);
   int i = 0;
   Node temp = L->front;
   while (temp != NULL){
      temp->index = i;
      i++;
      temp=temp->next;
   }
}

// deleteBack()
// Deletes element at back of L
// Pre: length(L) != 0
void deleteBack(List L){
   Node N = NULL;

   if( L==NULL ){
      printf("List Error: calling deleteBack() on NULL List reference\n");
      exit(1);
   }
   if( length(L)==0 ) {
      printf("List Error: calling deleteBack() on an empty List\n");
      exit(1);
   }
   N = L->back;
   if( length(L)>1 ) { L->back = L->back->prev; L->back->next=NULL; }
   else { L->front = L->back = NULL; }
   L->length--;
   freeNode(&N);
}
   
// delete()
// Deletes current element in List. Cursor is undefined after operation.
// Pre: length(L) > 0, getIndex(L) >= 0
void delete(List L){
   Node N = NULL;
   if( L==NULL ){
      printf("List Error: calling delete() on NULL List reference\n");
      exit(1);
   }
   if( length(L)==0 ){
      printf("List Error: calling delete() on an empty List\n");
      exit(1);
   }
   if( getIndex(L)<0 ){
      printf("List Error: calling delete() with undefined cursor\n");
      exit(1);
   } 
   if ( L->cursor == L->front ) deleteFront(L);
   else if ( L->cursor == L->back ) deleteBack(L);
   else {
      N = L->cursor;
      L->cursor->prev->next = L->cursor->next; 
      L->cursor->next->prev = L->cursor->prev;
      freeNode(&N);
      int i = 0;
      Node temp = L->front;
      while (temp != NULL){
         temp->index = i;
         i++;
         temp=temp->next;
      }
      L->length--;   
   }
   L->cursor = NULL;
}

// Other Functions ------------------------------------------------------------

// printList()
// Prints data elements in L on a single line to stdout.
void printList(FILE* out, List L){
   Node N = NULL;

   if( L==NULL ){
      printf("List Error: calling printList() on NULL List reference\n");
      exit(1);
   }
   for(N = L->front; N != NULL; N = N->next){
      fprintf(out, "%d ", N->data);
   }
   fprintf(out, "\n");
}

// copyList()
// Returns a new List representing the same integer sequence as List L. 
// The cursor in the new List is undefined, regardless of the state of the 
// cursor in L. L is unchanged.
List copyList(List L){
   List M = newList();
   Node N = L->front;
   while( N!=NULL ){
      append(M, N->data);
      N = N->next;
   }    
   return M;
}

File: pa3/List.java
// Spenser Estrada
// spmestra@unix.ucsc.edu 
// pa3

class List {
   private class Node {
      // fields
      Node prev;
      int index;
      Object data;
      Node next;
      // constructor
      Node (Object data) {
         this.prev = null;
         this.data = data;
         this.next = null;
      }
   }
   
   // fields
   private Node front;
   private Node back;
   private int length;
   private int index;
   private Node cursor;

   // constructor
   List() {
      front = back = null;
      length = 0;
      cursor = null;
      index = -1;
   }

   // Access Functions//////////
   
   // Returns the index of the cursor element in this list, or returns -1 
   // if the cursor element is undefined.
   int getIndex() {
      return index;
   }

   // Returns front element in this list
   // pre: length()>0
   Object front() {
      if(length() == 0) {
         throw new RuntimeException("List Error: front() called on empty list");
      }
      return front.data;
   }

   // Returns length of list
   int length() {
      return length;
   }
   
   // Returns back element in this list 
   Object back() {
      if(length() == 0) {
         throw new RuntimeException("List Error: back() called "
                                   +"on empty list");
      }
      return back.data;
   }

   // Returns cursor element in this list
   // Pre: length()>0, getIndex() >=0
   Object getElement() {
      if(length() == 0) {
         throw new RuntimeException("List Error: getElement() called"
                                   +" on empty list");
      }
      if(getIndex() < 0) {
         throw new RuntimeException("List Error: getElement() called" 
                                   +" without any item selected");
      }
      return cursor.data;
   }

   // Returns true if this List and L are same integer
   // sequence. The cursor is ignored in both lists.
   public boolean equals(Object x) {
      List L = null;
      boolean flag  = false;
      if( x instanceof List){
         Node N = this.front;
         Node M = L.front;
         flag = ( this.length() == L.length() ); 
         while ( flag && N!=null ) {
            flag = ( N.data == M.data );
            N = N.next;
            M = M.next;
         }
      }
      return flag;
   }

   // Manipulation procedures//////////

   // void clear(): resets this List to the empty state.
   void clear() {
      this.front = this.back = this.cursor = null;
      index = -1;
      this.length = 0;
   }

   // If 0<=i<=length()-1, moves cursor to the element, index i, otherwise 
   // the cursor becomes undefined.
   void moveTo(int i) {
      if ( 0 <= i && i <= length()-1) {
         cursor = front;
         index = 0;
         while (index != i) {
            cursor = cursor.next;
            index++;
         }
      } else {
         cursor = null;
         index = -1;
      }  
   }
   
   // If 0<getIndex()<=length()-1, moves the cursor one step toward the front 
   // of the list. If getIndex()==0, cursor becomes undefined.
   // If getIndex()==-1, cursor remains undefined. This operation is equivalent
   // to moveTo(getIndex()-1). 
   void movePrev() {
      if ( 0 < this.getIndex() && this.getIndex() <= length()-1 ) {
         cursor = cursor.prev;
         index--;
      } else {
         cursor = null;
         index = -1;
      }
   }

   // If 0<=getIndex()<length()-1, moves the cursor one step toward the back
   // of the list. If getIndex()==length-1, cursor becomes undefined. 
   // If index==-1, cursor remains undefined. This operation is equivalent 
   // to moveTo(getIndex()+1). 
   void moveNext () {
   if ( 0 <= this.getIndex() && this.getIndex() < length()-1 ) {
         cursor = cursor.next;
         index++;
      } else {
         cursor = null;
         index = -1;
      }
   }

   // Inserts new element before front element in list
   void prepend (Object data) {
      Node node = new Node(data);
      if( length() == 0 ) { 
         front = back = node;
      } else { 
         front.prev = node; node.next = front; front = node;
         if ( index>-1 ) index++; 
      }
      length++;
   }   

   // Inserts new element after back element in this List.
   void append(Object data) {
      Node node = new Node(data);
      if(length() == 0 ) { 
         front = back = node;
      } else {
         back.next = node; node.prev = back; back = node;  
      }
      length++;
   }

   // Inserts new element before cursor element in this List.
   // Pre: length()>0, getIndex()>=0
   void insertBefore(Object data) {
      if(length() == 0) {
         throw new RuntimeException("List Error: insertBefore(int data) called"
                                   +" on empty list");
      }
      if(this.getIndex() < 0) {
         throw new RuntimeException("List Error: insertBefore(int data) called"
                                   +" with undefined cursor");
      }
      Node node = new Node(data);
      if (cursor == front) {
         this.prepend(data);
      } else {
         cursor.prev.next = node; node.prev = cursor.prev; 
         node.next = cursor; cursor.prev = node;
         length++;
         index++; 
      } 
   }
   
   // Inserts new element after cursor element in this List.
   // Pre: length()>0, getIndex()>=0
   void insertAfter(Object data) {
      if(length() == 0) {
         throw new RuntimeException("List Error: insertAfter(int data) called"
                                   +" on empty list");
      }
      if(this.getIndex() < 0) {
         throw new RuntimeException("List Error: insertAfter(int data) called"
                                   +" with undefined cursor");
      }
      Node node = new Node(data);
      if (cursor == back) {
         this.append(data);
      } else {
         cursor.next.prev = node; node.next = cursor.next;
         node.prev = cursor; cursor.next = node;
         length++;
      }
   }  
   
   // Deletes the front element in this List.
   // Pre: length()>0
   void deleteFront() {
      if(length() == 0) {
         throw new RuntimeException("List Error: deleteFront() called on"
                                   +" empty list");
      }
      if ( cursor==front ) index = -1;
      else index--;
      if( length()==1 ) clear();
      else { front = front.next; front.prev = null;}
      length--;
   }

   // Deletes the back element in this List.
   // Pre: length()>0
   void deleteBack() {
      if(length() == 0) {
         throw new RuntimeException("List Error: deleteBack() called on" 
                                   +" empty list");
      }
      if (cursor==back )index = -1;
      if ( length()==1 ) clear();
      else { back.prev.next = null;  back.prev = back;}
      length--;
   }

   // Deletes current element in this List. Cursor is undefined after this 
   // operation
   // Pre: length() > 0, getIndex() >= 0.
   void delete() {
      if(length() == 0) {
         throw new RuntimeException("List Error: delete() called on"
                                   +" empty list");
      }
      if(getIndex() < 0) {
         throw new RuntimeException("List Error: delete() called with"
                                   +"undefine cursor");
      }
      if (cursor == front) this.deleteFront();
      else if (cursor == back) this.deleteBack();
      else {
         cursor.prev.next = cursor.next; cursor.next.prev = cursor.prev;
         int i = 0;
         for (Node temp = front; temp!= null; temp=temp.next) {
            temp.index = i;
            i++;
         }
         length--;   
      }
      cursor = null;
       index = -1;
   }
 
   // Other methods ///////////////
   
   // Overides Object's toString method. Creates a string
   // consisting of a space separated sequence of integers with front on the
   // left and back on the right. The cursor is ignored.
   public String toString() {
      String s = "";
      if (length() > 0){
         for (Node temp = front; temp != null; temp = temp.next) {
            s += (temp.data + " ");
         }
      }
      return s;
   }  
}

